ding/policy/bco.py

```diff
@@ -0,0 +1,191 @@
+import math
+import torch
+import torch.nn as nn
+import copy
+from torch.optim import SGD
+from torch.optim.lr_scheduler import LambdaLR
+from typing import List, Dict, Any, Tuple, Union, Optional
+from collections import namedtuple, deque
+from easydict import EasyDict
+from ding.policy import Policy
+from ding.model import model_wrap
+from ding.torch_utils import to_device
+from ding.utils import EasyTimer
+from ding.utils.data import default_collate, default_decollate
+from ding.rl_utils import q_nstep_td_data, q_nstep_sql_td_error, get_nstep_return_data, get_train_sample
+from ding.utils import POLICY_REGISTRY
+
+
+@POLICY_REGISTRY.register('bco')
+class BCOPolicy(Policy):
+    config = dict(
+        type='bc',
+        on_policy=False,
+        collect=dict(
+            # (int) Only one of [n_sample, n_step, n_episode] shoule be set
+            #n_sample=8,  # collect 8 samples and put them in collector
+            # (int) Cut trajectories into pieces with length "unroll_len".
+            unroll_len=1,
+        ),
+        other=dict(
+            # Epsilon greedy with decay.
+            eps=dict(
+                # (str) Decay type. Support ['exp', 'linear'].
+                type='exp',
+                start=0.95,
+                end=0.1,
+                # (int) Decay length(env step)
+                decay=10000,
+            ),
+            replay_buffer=dict(replay_buffer_size=10000, ),
+        ),
+    )
+
+    def _init_learn(self):
+        self._optimizer = SGD(
+            self._model.parameters(),
+            lr=self._cfg.learn.learning_rate,
+            weight_decay=self._cfg.learn.weight_decay,
+            momentum=0.9
+        )
+        self._timer = EasyTimer(cuda=True)
+
+        def lr_scheduler_fn(epoch):
+            if epoch <= self._cfg.learn.warmup_epoch:
+                return self._cfg.learn.warmup_lr / self._cfg.learn.learning_rate
+            else:
+                ratio = epoch // self._cfg.learn.decay_epoch
```

Comment:
`(epoch - self._cfg.learn.warmup_epoch) // self._cfg.learn.decay_epoch` can be better?